protocol/messenger_communities.go

```diff
@@ -1869,3 +1869,149 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+	}
+	response.CommunityChanges = []*communities.CommunityChanges{}
+
+	// We can safely assume that there's at least one community
+	// in the `MessengerResponse` because we've just created it
+	discordCommunity := response.Communities()[0]
+	addedCategoriesIds := make(map[string]string, 0)
+
+	for _, category := range discordCategories {
+		createCommunityCategoryRequest := &requests.CreateCommunityCategory{
+			CommunityID:  discordCommunity.ID(),
+			CategoryName: category.Name,
+			ChatIDs:      make([]string, 0),
+		}
+
+		r, err := m.CreateCommunityCategory(createCommunityCategoryRequest)
+		if err != nil {
+			return nil, err
+		}
+		response.AddCommunities(r.Communities())
+
+		// We can safely assume that there's at least one `CommunityChanges`
+		// in the `MessengerResponse` because we've just added a category
+		addedCategoryChange := r.CommunityChanges[0]
+		for _, addedCategory := range addedCategoryChange.CategoriesAdded {
+			addedCategoriesIds[category.ID] = addedCategory.CategoryId
+		}
+	}
+
+	messagesToSave := make([]*common.Message, 0)
+
+	for _, exportedChannel := range exportedData {
+
+		communityChat := &protobuf.CommunityChat{
+			Permissions: &protobuf.CommunityPermissions{
+				Access: protobuf.CommunityPermissions_NO_MEMBERSHIP,
+			},
+			Identity: &protobuf.ChatIdentity{
+				DisplayName: exportedChannel.Channel.Name,
+				Emoji:       "😎",
+				Description: exportedChannel.Channel.Description,
+			},
+			CategoryId: addedCategoriesIds[exportedChannel.Channel.CategoryID],
+		}
+
+		r, err := m.CreateCommunityChat(discordCommunity.ID(), communityChat)
+		if err != nil {
+			return nil, err
+		}
+		response.AddCommunities(r.Communities())
+
+		channelId := ""
+
+		// This looks like we keep overriding the `channelId` but in practice,
+		// this iterates only once
+		addedChatChange := r.CommunityChanges[0]
+		for chatId, _ := range addedChatChange.ChatsAdded {
+			channelId = chatId
+		}
+
+		for _, discordMessage := range exportedChannel.Messages {
+			exists, err := m.persistence.HasDiscordMessageAuthor(discordMessage.Author.GetId())
+			if err != nil {
+				return nil, err
+			}
+			if !exists {
+				err := m.persistence.SaveDiscordMessageAuthor(discordMessage.Author)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			payload, err := proto.Marshal(discordMessage)
+			if err != nil {
+				return nil, err
+			}
+
+			// Creating a `common.RawMessage` to get a `messageID`
+			rawMessage := &common.RawMessage{
+				Payload:        payload,
+				MessageType:    protobuf.ApplicationMetadataMessage_CHAT_MESSAGE,
+				Sender:         m.identity,
+				SkipEncryption: true,
+			}
+
+			messageID, err := m.sender.GetMessageID(rawMessage)
+			if err != nil {
+				return nil, err
+			}
+
+			layout := "2006-01-02T15:04:05+00:00"
+			timestamp, err := time.Parse(layout, discordMessage.Timestamp)
+			if err != nil {
+				return nil, err
+			}
+			discordMessage.Timestamp = fmt.Sprintf("%d", timestamp.Unix())
```

Comment:
Oh and here I'm changing it to actually become that unix timestamp as string. Makes it easier to deal with later.